### ISSUE TITLE
Two emerging patterns for specifying splits in `brainsets.datasets`

### ISSUE THREAD

**Issue by vinamarora8** [neuro-galaxy/torch_brain#323]
There seem to be two patterns emerging when it comes to how `torch_brain.datasets` let users to specify train/val/test splits in `Dataset` instantiation:

Type1. Split specified during `get_sampling_intervals`:
 ```python
ds = MyDataset(...)
train_intervals = ds.get_sampling_intervals("train")
val_intervals = ds.get_sampling_intervals("valid")
test_intervals = ds.get_sampling_intervals("test")
```

Type2. Split specified during Dataset instantiation:
```python
train_ds = MyDataset(..., split="train")
train_intervals = train_ds.get_sampling_intervals()

val_ds = MyDataset(..., split="val")
val_intervals = val_ds.get_sampling_intervals()

test_ds = MyDataset(..., split="test")
test_intervals = test_ds.get_sampling_intervals()
```

All dataset implementations at the moment fall under Type1, except `Neuroprobe2025`, which is implemented as Type2. I also see Type2 being chosen in a (not public at the moment) benchmarking project @AlexandreAndr & @ShivashriganeshMahato  are developing. In both cases, Type1 is insufficient to achieve the functionality needed by the underlying tasks.

**My question is:** Do we allow both patterns in the long term? Or should we enforce following a single pattern?

A strong reason to follow a single pattern would be `NestedDataset`. It expects a single interface for `get_sampling_intervals()` from all of its underlying datasets. In other words, `<NestedDataset>.get_sampling_intervals("train")` would fail if one of its nested Datasets fails at `get_sampling_intervals("train")`.

If we were to enforce a single pattern, I would be leaning towards Type2 for the following reasons:

- It is the more flexible pattern and provides more power to the Dataset constructor. What you can do with Type1, you can also do with Type2, but not the other way around. Neuroprobe2025 is an example where we needed that extra flexibility, and it _could not_ have used Type1. It sets the recording_ids etc based on the value of `split`, e.g. when operating in cross-session regime.

- In case users are using random augmentation transforms (e.g. UnitDropout), they would anyway need to create separate `train_ds`, `val_ds`. So the benefit of instantiating a single Dataset, and using it for all splits is not a universal advantage.

Would love thoughts!

CC: @milosobral @AlexandreAndr

**Comment by AlexandreAndr** [comment]
Great questions!
About the single pattern enforcement, it could be nice to have some flexibility in general but the `NestedDataset` point made me think that it might be better to move for single pattern.
If so, type 2 design seems the best for me too and mainly for the first reason you mentioned. 
In general, avoiding sharing/leaking information between splits is something we want to ensure. Some splits can requires differences at the dataset level, for example different recording ids (as you mentionned), a fixed set of hyperparameters or some data attribute only accessible on the test dataset and removed in the others.

**Comment by milosobral** [comment]
I would tend to agree with both points:
- I think enforcing single pattern is important for NestedDataset reason but also consistency
- Agree that Type 2 probably makes more sense
@vinamarora8 When you say enforce, are you thinking about adding a strict protocol/interface or do you mean just making sure we follow that convention in the code?

**Comment by vinamarora8** [comment]
Great! Looks like we are all aligned.

> When you say enforce, are you thinking about adding a strict protocol/interface or do you mean just making sure we follow that convention in the code?

@milosobral I mean the latter. Just ensure consistency in all Dataset subclasses that we provide in our pkg.
